File: sdks/undolog-py/undolog_sdk/signature.py
"""Deterministic canonical JSON and BLAKE3 call-signature computation.

This module is the most important part of the SDK. It MUST produce byte-for-byte
identical output to the Rust crate ``undolog-types`` for the ``canonical_json``
and ``CallSignature::compute`` functions.

Cross-language invariant:
    Same inputs → same 64-char hex signature, regardless of language or platform.
"""

from __future__ import annotations

import json
import math
import struct
import uuid
from typing import Any

_ES6_EXP_THRESHOLD = 21


def _es6_float(value: float) -> str:
    """Serialise a float exactly as ECMAScript ``JSON.stringify`` does.

    Python's ``json.dumps`` and ``repr`` use different formatting rules than
    ECMAScript (e.g. ``1e-06`` vs ``0.000001``, ``-0.0`` vs ``0``). This
    function reproduces the ECMAScript Number::toString output so the canonical
    JSON bytes match the TypeScript SDK byte-for-byte.

    Args:
        value: A finite float.

    Returns:
        The ECMAScript-compatible serialisation of the float.

    Raises:
        ValueError: If ``value`` is not finite (``nan``, ``inf``, ``-inf``).
    """
    if not math.isfinite(value):
        raise ValueError(f"Invalid JSON number: {value!r}")
    if value == 0:
        return "0"
    text = repr(value)
    sign = "-" if text.startswith("-") else ""
    if sign:
        text = text[1:]
    exp_str = ""
    exp_val = 0
    q = text.find("e")
    if q > 0:
        exp_str = text[q:]
        if exp_str[2:3] == "0":
            exp_str = exp_str[:2] + exp_str[3:]
        text = text[:q]
        exp_val = int(exp_str[1:])
    first, dot, last = text.partition(".")
    if last == "0":
        dot = ""
        last = ""
    if 0 < exp_val < _ES6_EXP_THRESHOLD:
        first += last
        last = ""
        dot = ""
        exp_str = ""
        pad = exp_val + 1 - len(first)
        if pad > 0:
            first += "0" * pad
    elif -7 < exp_val < 0:
        last = first + last
        first = "0"
        dot = "."
        exp_str = ""
        pad = -exp_val - 1
        if pad > 0:
            last = "0" * pad + last
    return f"{sign}{first}{dot}{last}{exp_str}"


def canonical_json(value: Any) -> str:
    """Produce a deterministic, sorted-key JSON string suitable for hashing.

    serde_json serialises mappings in insertion order which varies across
    languages. This function recursively sorts all keys so that
    ``{"b":1,"a":2}`` and ``{"a":2,"b":1}`` produce the same canonical string.

    Numbers follow the ECMAScript ``JSON.stringify`` rules (RFC 8785):
    ``0.0`` and ``-0.0`` both serialise as ``0``, small and large values use
    exponential notation matching JavaScript, and exponents have no leading
    zeros. This keeps the canonical bytes identical to the TypeScript and Rust
    SDKs.

    Args:
        value: A JSON-compatible Python object (dict, list, str, int, float,
            bool, None).

    Returns:
        Compact JSON string with recursively sorted keys, no whitespace,
        matching the TypeScript and Rust SDK output for leaf values.

    Raises:
        ValueError: If ``value`` contains a non-finite float (``nan``,
            ``inf``, ``-inf``). These are rejected to match the TypeScript
            SDK, which throws a ``TypeError`` for the same inputs.
    """
    if isinstance(value, dict):
        pairs = [(k, canonical_json(v)) for k, v in value.items()]
        pairs.sort(key=lambda x: x[0])
        inner = ",".join(f'"{k}":{v}' for k, v in pairs)
        return f"{{{inner}}}"
    if isinstance(value, (list, tuple)):
        inner = ",".join(canonical_json(v) for v in value)
        return f"[{inner}]"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, float):
        return _es6_float(value)
    if isinstance(value, int):
        return str(value)
    if value is None:
        return "null"
    return _escape_json_string(value)


def _escape_json_string(value: str) -> str:
    """Encode a string as JSON with ``ensure_ascii`` semantics.

    Matches Python ``json.dumps(value, ensure_ascii=True)`` output.

    Args:
        value: String to encode.

    Returns:
        The JSON-encoded string literal (including surrounding double quotes).
    """
    return json.dumps(value, ensure_ascii=True)


def call_signature(
    session_id: str,
    step_index: int,
    tool_name: str,
    args: Any,
) -> str:
    """Compute the canonical call signature for a tool call.

    Every SDK (Rust, Python, TypeScript) MUST produce the same 64-character
    lowercase hex output for the same inputs. The length-prefixed encoding
    prevents boundary attacks where two different (name, args) pairs could
    produce the same byte sequence without delimiters.

    The BLAKE3 hash is computed over the following byte stream:

        [session_id: 16 bytes]
        [step_index: 4 bytes LE]
        [tool_name_len: 4 bytes LE][tool_name: N bytes UTF-8]
        [canonical_args_len: 4 bytes LE][canonical_args: M bytes UTF-8]

    Args:
        session_id: UUID string identifying the session.
        step_index: Monotonically increasing step counter within the session.
        tool_name: Logical name of the tool being called.
        args: JSON-compatible object (dict, list, etc.) representing the
            tool arguments. Will be canonicalized before hashing.

    Returns:
        64-character lowercase hex string (BLAKE3-256).

    Raises:
        ValueError: If session_id is not a valid UUID.
    """
    try:
        sid = uuid.UUID(session_id)
    except (ValueError, AttributeError) as exc:
        raise ValueError(
            f"session_id must be a valid UUID string, got {session_id!r}"
        ) from exc

    canon = canonical_json(args)
    name_bytes = tool_name.encode("utf-8")
    args_bytes = canon.encode("utf-8")

    # blake3 is imported lazily: the C-ext wheel is not always installed in
    # every environment, and this module is the only one that needs it.
    import blake3

    hasher = blake3.blake3()

    # 1. session_id as 16 raw bytes, network byte order (fixed width → no prefix)
    hasher.update(sid.bytes)

    # 2. step_index as 4-byte little-endian
    hasher.update(struct.pack("<I", step_index))

    # 3. length-prefixed tool_name
    hasher.update(struct.pack("<I", len(name_bytes)))
    hasher.update(name_bytes)

    # 4. length-prefixed canonical args JSON
    hasher.update(struct.pack("<I", len(args_bytes)))
    hasher.update(args_bytes)

    return hasher.hexdigest()

File: sdks/undolog-ts/src/signature.ts
/** Deterministic canonical JSON and BLAKE3 call-signature computation.
 *
 * This is the most critical module in the SDK. It MUST produce byte-for-byte
 * identical output to the Python SDK's ``undolog_sdk.signature`` module for the
 * same inputs: same inputs, same 64-char hex signature, regardless of language
 * or platform.
 *
 * @module
 */

import { blake3 } from "@noble/hashes/blake3.js";
import { bytesToHex } from "@noble/hashes/utils.js";

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

/** Escape a string for JSON with Python's ``ensure_ascii=True`` semantics.
 *
 * Non-ASCII characters (code point >= 0x80) are emitted as ``\\uXXXX`` escape
 * sequences. This matches Python ``json.dumps(..., ensure_ascii=True)`` and
 * Rust ``serde_json`` output.
 *
 * @param s - Raw string to escape.
 * @returns JSON-encoded string literal (including surrounding double quotes).
 */
function escapeJsonString(s: string): string {
  let result = '"';
  for (let i = 0; i < s.length; i++) {
    const ch = s.charAt(i);
    const code = s.charCodeAt(i);
    switch (ch) {
      case '"':
        result += '\\"';
        break;
      case "\\":
        result += "\\\\";
        break;
      case "\b":
        result += "\\b";
        break;
      case "\f":
        result += "\\f";
        break;
      case "\n":
        result += "\\n";
        break;
      case "\r":
        result += "\\r";
        break;
      case "\t":
        result += "\\t";
        break;
      default:
        if (code < 0x20) {
          result += `\\u${code.toString(16).padStart(4, "0")}`;
        } else if (code < 0x80) {
          result += ch;
        } else {
          result += `\\u${code.toString(16).padStart(4, "0")}`;
        }
    }
  }
  return `${result}"`;
}

/** Produce a deterministic, sorted-key JSON string suitable for hashing.
 *
 * ``JSON.stringify`` serialises object properties in insertion order, which
 * varies across languages. This function recursively sorts all keys so that
 * ``{"b":1,"a":2}`` and ``{"a":2,"b":1}`` produce the same canonical string.
 *
 * Non-ASCII characters are escaped (``ensure_ascii`` semantics) to match
 * Python's ``json.dumps(..., ensure_ascii=True)`` and Rust ``serde_json``.
 *
 * @param value - A JSON-compatible value (object, array, string, number,
 *   boolean, null).
 * @returns Compact JSON string with recursively sorted keys, no whitespace.
 *
 * @example
 * ```ts
 * canonicalJson({ b: 1, a: 2 }); // '{"a":2,"b":1}'
 * canonicalJson("hello");        // '"hello"'
 * ```
 */
export function canonicalJson(value: unknown): string {
  if (value === null) {
    return "null";
  }
  if (typeof value === "boolean") {
    return value ? "true" : "false";
  }
  if (typeof value === "number") {
    return JSON.stringify(value);
  }
  if (typeof value === "string") {
    return escapeJsonString(value);
  }
  if (Array.isArray(value)) {
    const items = value.map((v) => canonicalJson(v));
    return `[${items.join(",")}]`;
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    const pairs = keys.map((k) => `${escapeJsonString(k)}:${canonicalJson(obj[k])}`);
    return `{${pairs.join(",")}}`;
  }
  throw new TypeError(`Cannot serialise value of type ${typeof value}`);
}

/** Encode a 32-bit unsigned integer as 4 bytes in little-endian order.
 *
 * @param n - Integer value (will be truncated to 32 bits).
 * @returns A 4-byte Uint8Array in little-endian byte order.
 */
function le32(n: number): Uint8Array {
  const buf = new Uint8Array(4);
  buf[0] = n & 0xff;
  buf[1] = (n >> 8) & 0xff;
  buf[2] = (n >> 16) & 0xff;
  buf[3] = (n >> 24) & 0xff;
  return buf;
}

/** Parse a UUID string into its 16 raw bytes (network byte order).
 *
 * This matches Python's ``uuid.UUID(s).bytes`` representation: the hex groups
 * are concatenated and decoded in big-endian byte order.
 *
 * @param uuid - UUID string in standard ``xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx``
 *   format.
 * @returns 16-byte Uint8Array.
 * @throws {TypeError} If the UUID string is not valid.
 */
function parseUuidBytes(uuid: string): Uint8Array {
  if (!UUID_RE.test(uuid)) {
    throw new TypeError(`Invalid UUID string: ${uuid}`);
  }
  const hex = uuid.replace(/-/g, "");
  const bytes = new Uint8Array(16);
  for (let i = 0; i < 16; i++) {
    bytes[i] = Number.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
  }
  return bytes;
}

/** Compute the canonical BLAKE3 call signature for a tool call.
 *
 * Every SDK (Rust, Python, TypeScript, C#) MUST produce the same 64-character
 * lowercase hex output for the same inputs. The length-prefixed encoding
 * prevents boundary attacks where two different (name, args) pairs could
 * produce the same byte sequence without delimiters.
 *
 * The BLAKE3 hash is computed over the following byte stream:
 *
 * - ``[session_id: 16 bytes]``
 * - ``[step_index: 4 bytes LE]``
 * - ``[tool_name_len: 4 bytes LE][tool_name: N bytes UTF-8]``
 * - ``[canonical_args_len: 4 bytes LE][canonical_args: M bytes UTF-8]``
 *
 * @param sessionId - UUID string identifying the session.
 * @param stepIndex - Monotonically increasing step counter within the session.
 * @param toolName - Logical name of the tool being called.
 * @param args - JSON-compatible object (dict, list, etc.) representing the
 *   tool arguments. Will be canonicalised before hashing.
 * @returns 64-character lowercase hex string (BLAKE3-256).
 * @throws {TypeError} If ``sessionId`` is not a valid UUID.
 *
 * @example
 * ```ts
 * const sig = callSignature(
 *   "550e8400-e29b-41d4-a716-446655440000",
 *   0,
 *   "send_email",
 *   { to: "alice@example.com" },
 * );
 * console.log(sig); // 64-character hex string
 * ```
 */
export function callSignature(
  sessionId: string,
  stepIndex: number,
  toolName: string,
  args: unknown,
): string {
  const sidBytes = parseUuidBytes(sessionId);
  const canon = canonicalJson(args);
  const nameBytes = new TextEncoder().encode(toolName);
  const argsBytes = new TextEncoder().encode(canon);

  const hasher = blake3.create();

  // 1. session_id as 16 raw bytes, network byte order (fixed width, no prefix)
  hasher.update(sidBytes);

  // 2. step_index as 4-byte little-endian
  hasher.update(le32(stepIndex >>> 0));

  // 3. length-prefixed tool_name
  hasher.update(le32(nameBytes.length));
  hasher.update(nameBytes);

  // 4. length-prefixed canonical args JSON
  hasher.update(le32(argsBytes.length));
  hasher.update(argsBytes);

  return bytesToHex(hasher.digest());
}
